datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -1125,6 +1126,14 @@ impl LogicalPlan {
         }
     }
 
+    /// These are invariants to hold true for each logical plan.
+    pub fn assert_invariants(&self) -> Result<()> {
+        // Refer to <https://datafusion.apache.org/contributor-guide/specification/invariants.html#relation-name-tuples-in-logical-fields-and-logical-columns-are-unique>
+        assert_unique_field_names(self)?;
+
```

Comment:
I removed the `assert_unions_are_valid` because I was not sure if it should be a LP invariant (it's [not listed here](https://datafusion.apache.org/contributor-guide/specification/invariants.html#id4)), and it caused ~17 test failures.

Reply:
Good call -- I think we could/should discover/enforce new invariants as follow on PRs after we have sorted out this basic API